pkg: finder -> resourcemonitor

ffromani opened the pull request:
This PR drops the existing pkg/finder and replaces with
pkt/resourcemonitor. The functionality of the package is preserved with
minimal API change, and no change of behaviour is intended for
`nfd-topology-updater`.

Rationale for this rewrite is that `pkg/finder` mostly grow up over
time with no organic design and as aggregation of feature.
Thus, the design was messy and the code was hard to follow and to test.

The current rewrite wants to preserve the responsabilities and
the basic design tenets of the package, but aims to provide
a much cleaner interface, better readability and improved testability.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented on pkg/topologyinfo/cpus/cpus.go:
```diff
@@ -0,0 +1,124 @@
+import (
+	"strconv"
+
+	"github.com/fromanirh/topologyinfo/sysfs"
```
wrong import, will fix in the next upload.

ffromani replied:
fixed!

cynepco3hahue commented:
@fromanirh Does it WIP? Because I can not see the implementation for
```
nodes, err := numa.NewNodesFromSysFS(sysfsPath)
if err != nil {
    return nil, err
}

dists, err := distances.NewDistancesFromSysfs(sysfsPath)
if err != nil {
	return nil, err
}
```

ffromani commented:
> @fromanirh Does it WIP? Because I can not see the implementation for
> 
> ```
> nodes, err := numa.NewNodesFromSysFS(sysfsPath)
> if err != nil {
>     return nil, err
> }
> 
> dists, err := distances.NewDistancesFromSysfs(sysfsPath)
> if err != nil {
> 	return nil, err
> }
> ```

Already merged on a previous PR: https://github.com/swatisehgal/node-feature-discovery/tree/rte/pkg/topologyinfo/numa

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,243 @@
+		contDevs = append(contDevs, dev)
+	}
+
+	cpuRes := make(map[int][]string)
```
I think this should be named cpuResPerNuma or on those lines to be more descriptive

ffromani replied:
good point, done

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,243 @@
+
+	return contDevs
+}
+
```
Lets add a comment providing a brief description of what this function does

ffromani replied:
sure, done.

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,243 @@
+	return fmt.Sprintf("node-%d", numaID)
+}
+
+func MakeNodeCapacity(devices []*podresourcesapi.ContainerDevices) map[int]map[v1.ResourceName]int64 {
```
Same as above

ffromani replied:
yep, done

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,243 @@
+
+}
+
+func makeResourceMap(numaNodes int, devices []*podresourcesapi.ContainerDevices) map[string]map[string]int {
```
Same

ffromani replied:
done!

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,243 @@
+	return resourceMap
+}
+
+func makeCostsPerNumaNode(numaInfo *numa.Nodes, numaDists *distances.Distances, numaIDSrc int) (map[string]int, error) {
```
Here too

ffromani replied:
done!

swatisehgal commented on pkg/topologyinfo/cpus/cpus.go:
```diff
@@ -0,0 +1,124 @@
+	"sigs.k8s.io/node-feature-discovery/pkg/topologyinfo/sysfs"
+)
+
+// CPUIdList is a list of CPU IDs (integer core identifier)
```
This is great, it would be great to have description like this for other data structures in type.go too

ffromani replied:
done!

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,243 @@
+
+		costs, err := makeCostsPerNumaNode(nodeResourceData.numaInfo, nodeResourceData.numaDists, nodeNum)
+		if err != nil {
+			// TODO log and move forward
```
Lets add a log message here

ffromani replied:
fixed

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,243 @@
+	for _, cpuID := range availRes.GetCpuIds() {
+		nodeID, ok := cpus.GetNodeIDForCPU(int(cpuID))
+		if !ok {
+			// TODO log this is weird
```
Lets add a log here

ffromani replied:
done

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,243 @@
+
+	return &nodeResourceInstance, nil
+}
+
```
Lets add a comment providing a brief description of what this function does

ffromani replied:
done

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,243 @@
+	}
+	return zones
+}
+
```
Lets add a comment providing a brief description of what this function does

ffromani replied:
done

swatisehgal reviewed (changes requested):
Looks good overall Just a few minor comments:

- Log messages need to be added at a couple of places

- Add comments describing methods and data structure for better code readability

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,257 @@
+	return &nodeResourceInstance, nil
+}
+
+// Aggregate prodices the mapping (numa zone name) -> Zone from the given PodResources.
```
typo: provides

ffromani replied:
thanks, fixed

swatisehgal replied:
prodices :) ?

ffromani replied:
ops :) fixing in the next upload

ffromani commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -0,0 +1,257 @@
+
+type nodeResources struct {
+	perNUMACapacity map[int]map[v1.ResourceName]int64
+	// we should use a different map for each resource type: resourceName -> resourceID -> nodeID
```
stale comment, should read like
```
mapping: resourceName -> resourceID -> nodeID
```

ffromani replied:
fixed

swatisehgal commented on pkg/topologyinfo/cpus/cpus.go:
```diff
@@ -0,0 +1,126 @@
+ * Copyright 2020 Red Hat, Inc.
+ */
+
+package cpus
```
I noticed [here](https://github.com/fromanirh/topologyinfo/tree/main/cpus) that you already worked on cpu_test.go. If the implementation is exactly the same, can we add the cpu_test.go as part of this PR too? Just a suggestion/reminder :)

ffromani replied:
This is exactly where I am expecting the discussion to happen. `topologyinfo` here is indeed a narrowed-down, copy/pasted (poor's man vendor) clone of https://github.com/fromanirh/topologyinfo/. The problem is exactly this: being a copy/paste we do have a test suite, but it is not integrated at all with the rest of the tests found elsewhere in the codebase.

I have no preference regarding how to address this issue (if at all)

swatisehgal replied:
Keep me honest, but I thought that [this](https://github.com/swatisehgal/node-feature-discovery/blob/master/Makefile#L103) would take care of executing the tests?

ffromani replied:
absolutely. But please compare [this](https://github.com/swatisehgal/node-feature-discovery/blob/rte/pkg/topologyinfo/numa/numa_test.go) with [this](https://github.com/swatisehgal/node-feature-discovery/blob/rte/pkg/nfd-worker/nfd-worker-internal_test.go).

We can clearly see the tests I added on `topologyinfo` are written with different style, conventions, tools (`sysfs/fake` instead of `mock`s). This is indeed caused by the poor man's vendoring approach I mentioned. Nevertheless `pkg/topologyinfo` feels alien compared to the rest of the codebase.

Do we want to take care of this? and if so, how?

swatisehgal replied:
I see your point. I am open to suggestions on the best way to move ahead. I see a couple of options here:

1. Add the topologyinfo pkg code in this repo for now (merge this PR) and later add tests that are of the same style as that in the repo (as another PR).  

1. Import pkg [https://github.com/fromanirh/topologyinfo](https://github.com/fromanirh/topologyinfo) now, later update the import path to https://github.com/openshift-kni/topologyinfo

ffromani replied:
I think the latter is better because makes easier for us to pivot to other information-gathering packages (e.g. `ghw`) later on if we feel it's better or if NFD maintainers ask for it. Updating accordingly.

swatisehgal reviewed (commented):
A small typo but other than that we are good to go!

ffromani committed:
pkg: finder -> resourcemonitor

This PR drops the existing pkg/finder and replaces with
pkt/resourcemonitor. The functionality of the package is preserved with
minimal API change, and no change of behaviour is intended for
`nfd-topology-updater`.

Rationale for this rewrite is that `pkg/finder` mostly grow up over
time with no organic design and as aggregation of feature.
Thus, the design was messy and the code was hard to follow and to test.

The current rewrite wants to preserve the responsabilities and
the basic design tenets of the package, but aims to provide
a much cleaner interface, better readability and improved testability.

Signed-off-by: Francesco Romani <fromani@redhat.com>

swatisehgal commented:
LGTM! Thanks for your great work on this PR :)
